src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.transaction;
+
+import java.util.ArrayList;
+import java.util.Objects;
+import net.neoforged.neoforge.transfer.ResourceHandlerDeprecationHandling;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A journal that modifies itself during transactions,
+ * saving snapshots of its state in objects of type {@code T} in case it needs to revert to a previous state.
+ *
+ * <h3>How to use from subclasses</h3>
+ * <ul>
+ * <li>Call {@link #updateSnapshots} right before the state of your subclass is modified in a transaction.</li>
+ * <li>Override {@link #createSnapshot}: it is called when necessary to create an object representing the state of your subclass.</li>
+ * <li>Override {@link #revertToSnapshot}: it is called when necessary to revert to a previous state of your subclass.</li>
+ * <li>You may optionally override {@link #onCommit}: it is called at the end of a transaction that modified the state.
+ * For example, it could contain a call to {@code setChanged()}.</li>
+ * <li>(Advanced!) You may optionally override {@link #releaseSnapshot}: it is called once a snapshot object will not be used,
+ * for example you may wish to pool expensive state objects.</li>
+ * </ul>
+ *
+ * <h3>More technical explanation</h3>
+ *
+ * <p>{@link #updateSnapshots} should be called before any modification.
+ * This will save the state of this participant using {@link #createSnapshot} if no state was already saved for that transaction.
+ * When the transaction is aborted and changes need to be rolled back, {@link #revertToSnapshot} will be called
+ * to signal that the current state should revert to that of the snapshot.
+ * The snapshot object is then {@linkplain #releaseSnapshot released}, and can be cached for subsequent use, or discarded.
+ *
+ * <p>When an outer transaction is committed, {@link #revertToSnapshot} will not be called so that the current state of this participant
+ * is retained. {@link #onCommit} will be called after the transaction is closed
+ * and then {@link #releaseSnapshot} will be called because the snapshot is not necessary anymore.
+ *
+ * @param <T> The objects that this participant uses to save its state snapshots.
+ */
+public abstract class SnapshotJournal<T> implements Transaction.CloseCallback, TransactionContext.RootCloseCallback {
```

Comment:
If we want to we can make this support `null` snapshots by masking nulls. This would also simplify the cases where we don't actually need the snapshot, and `Void` would be the most appropriate type. I can send a PR to this PR.